src/charm.py

```diff
@@ -0,0 +1,257 @@
+        self,
+    ) -> tuple[LibraryWriteBehaviour, ops.ActiveStatus] | tuple[None, ops.BlockedStatus]:
+        library_write_behaviour = self.model.config[LIBRARY_WRITE_CONFIG]
+        if library_write_behaviour not in LIBRARY_WRITE_BEHAVIOURS:
+            msg = f"invalid {LIBRARY_WRITE_CONFIG}: '{library_write_behaviour}'"
+            return None, ops.BlockedStatus(msg)
+        return cast(LibraryWriteBehaviour, library_write_behaviour), ops.ActiveStatus()
+
+    def _push_and_extract_library(self, container: ops.Container, library: bytes) -> None:
+        # use dtrx to extract library due to its consistent behaviour over many archive types
+        # could just add dtrx to the docker image? but it's somehow nice to use the off the shelf image
+        try:
+            container.exec(['which', 'dtrx']).wait()
+        except ops.pebble.ExecError:
+            logger.debug('_push_and_extract_library: installing dependencies')
+            container.exec(['apt', 'update']).wait()
+            container.exec(['apt', 'install', 'dtrx', '-y']).wait()
+        logger.debug('_push_and_extract_library: copying library')
+        container.push('/books/library.zip', library)
+        logger.debug('_push_and_extract_library: extracting library')
+        container.exec(
+            ['dtrx', '--noninteractive', '--overwrite', 'library.zip'],
+            working_dir='/books',
+        ).wait()
+        # dtrx always extracts into a folder named after the archive
+        logger.debug('_push_and_extract_library: flattening /books/library/ to /books/')
+        self._move_directory_contents_to_parent(container, directory='/books/library')
+        container.remove_path('/books/library.zip')
+        # library.zip may contain the library contents directly, or inside a 'Calibre Library' directory
+        if container.exists('/books/Calibre Library'):
+            logger.debug(
+                '_push_and_extract_library: flattening /books/Calibre Library/ contents to /books'
+            )
+            self._move_directory_contents_to_parent(container, directory='/books/Calibre Library')
+        logger.debug('push_and_extract_library: done')
+
+    def _move_directory_contents_to_parent(
+        self, container: ops.Container, directory: Path | str
+    ) -> None:
+        directory = str(directory)
+        move_contents_up_one_level = [
+            'bash',
+            '-c',
+            (
+                'shopt -s nullglob ; '
+                'mv --force --target-directory=../ ./* ./.[!.]* '
+                '|| true'  # ignore no matches in glob ... in addition to other errors >:(
+                # mv will fail if any directories colide (--force doesn't overwrite dirs)
+                # but they shouldn't since we only run this if we first clean /books/
+            ),
+        ]
+        container.exec(move_contents_up_one_level, working_dir=directory).wait()
+        container.exec(['rmdir', directory]).wait()  # error if not empty
```

Comment:
Could this be `remove_path`?

Reply:
It could be. During development I liked that `rmdir` would error if the directory wasn't empty, which helped me catch some issues that would have been difficult to debug otherwise, but now that everything is working, do you think it be better to switch to using `remove_path`?

Reply:
Using recursive=False would get you those errors still, I believe. It seems the more 'native' way to do it, but exec is ok.

Reply:
Oh yeah that's ideal, thanks